File: packages/core/src/digest.ts
import { createHash } from 'node:crypto';
import { Bundle, PATHS } from './bundle.js';

/**
 * Deterministic content digest over preserved file bytes, excluding signature.json.
 * Independent of archive format and insertion order, so unpack(pack(b)) reproduces it.
 */
export function canonicalDigest(bundle: Bundle): string {
  const entries = bundle
    .entries()
    .filter(([path]) => path !== PATHS.signature)
    .sort(([a], [b]) => (a < b ? -1 : a > b ? 1 : 0));

  let canonical = '';
  for (const [path, bytes] of entries) {
    const fileHash = createHash('sha256').update(bytes).digest('hex');
    canonical += `${path}\n${fileHash}\n`;
  }
  return createHash('sha256').update(canonical, 'utf8').digest('hex');
}

File: apps/studio/src-tauri/src/core/digest.rs
use sha2::{Digest, Sha256};

use super::bundle::{paths, Bundle};

fn sha256_hex(bytes: &[u8]) -> String {
    let mut h = Sha256::new();
    h.update(bytes);
    hex_encode(&h.finalize())
}

fn hex_encode(bytes: &[u8]) -> String {
    bytes.iter().map(|b| format!("{b:02x}")).collect()
}

/// Port of packages/core/src/digest.ts — sha256 over sorted `path\n<file sha256 hex>\n` lines,
/// signature.json excluded. Archive-independent by construction.
pub fn canonical_digest(bundle: &Bundle) -> String {
    let mut canonical = String::new();
    for (path, bytes) in bundle.entries() {
        if path == paths::SIGNATURE {
            continue;
        }
        canonical.push_str(path);
        canonical.push('\n');
        canonical.push_str(&sha256_hex(bytes));
        canonical.push('\n');
    }
    sha256_hex(canonical.as_bytes())
}
